### DISCUSSION TITLE
Grpc and quarkus

### DISCUSSION THREAD

**Discussion by Casillas0905** [quarkusio/quarkus#40266] [Q&A]
HI 
In the new update of quarkus (3.9.4), has been any changes related to grpc services?
I'm trying to update to quarkus 3.9.4 from 3.2.10.Final, and i'm getting errors for the grpc service, this it what it says: io.grpc.StatusRuntimeException: UNAVAILABLE: io exception.
If anyone know anything i would appreciate it

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
There are many changes - most of them related to the new `protoc`. Do you have a reproducer for your specific issue?

**Reply by Casillas0905** [reply]
Do you have some documentation about?, because i think to reproduce the error its going to take a while because its company code, so i have to redo it

**Reply by cescoffier** [reply]
Unfortunately, no, we have been fighting lots of change with the new protoc (new system library required, change of behavior of some parameter).

**Reply by Casillas0905** [reply]
ok , then i'll be beack when i have a reproducer, thanks anyway

**Reply by Casillas0905** [reply]
Maybe this help you understand the problem, this are the logs before and after the update, in the second one, is missing the grpc server feature:
Installed features: [bgcore-extension, cdi, grpc-client, grpc-server, logging-json, micrometer, oidc, resteasy-reactive, security, smallrye-context-propagation, smallrye-health, vertx]
Installed features: [bgcore-extension, cdi, grpc-client, logging-json, micrometer, oidc, rest, security, smallrye-context-propagation, smallrye-health, vertx]

I also add this dependency in the pom file but it did nothing: <dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-grpc</artifactId>
    <version>${quarkus.version}</version>
</dependency>


Maybe there is new dependency i have to add or something similar.

**Reply by cescoffier** [reply]
That means that it cannot find the gRPC service. How do you generate the Java classes from `proto`? Is the class annotated with `@GrpcService` in your project?

**Reply by Emanuel-Rad** [reply]
Hi @cescoffier,

I'm @Casillas0905 's colleague. The issue seems to be that during testing, the gRPC client is trying to call on the default 9001 port, instead of the one specified. The setup is as follows:

(1) we have a docker container running with our `@GrpcService MyService`
(2) in a separate library, we have this class that should make a gRPC call to (1):
```
public class MyHandler {

    @GrpcClient
    MyService myService;

    public boolean call(...)
    {
        return myService.call(...)..await()...;
    }
}
```
(3) The test setup is as follows:
- the docker container (1) is started using [testcontainers](https://testcontainers.com/), that means that the exposed port is random - let's say 65315
- The `quarkus.grpc.clients.myService.port` property is set to 65315 using `System.setProperty()` in a `QuarkusUnitTest`
- The `MyHandler` from (2) is injected and called
- The call fails because it's trying to contact the docker container (1) on the default port 9001, instead of 65315
Here's some snippets of the test:

```
public class MyTest {

    public static final GenericContainer myGrpcServerContainer = new GenericContainer(DockerImageName.parse("XXX:latest"))
            .withExposedPorts(5000); // we remapped the default 9000 gRPC port to 5000

    @RegisterExtension
    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
        .setBeforeAllCustomizer(() -> {
            myGrpcServerContainer.start();
            String port= myGrpcServerContainer.getMappedPort(5000).toString();   // this returns the 65315
            System.setProperty("quarkus.grpc.clients.myService.port", port);
        })
        .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                .addAsResource("mytest.properties", "application.properties");
        );

    @Inject
    MyHandler myHandler;

    @ParameterizedTest
    @ValueSource(strings = { "Foo", "Bar", "Invalid", "" })
    void MyHandlerTest(String role)
    {
        assertEquals(true, myHandler.call(...));
    }
}
```

**Reply by cescoffier** [reply]
Is MyHandler a gRPC client? Can you show the application config?

**Reply by cescoffier** [reply]
Also, was this approach working before?

**Reply by Emanuel-Rad** [reply]
MyHandler is just an ApplicationScoped bean, a wrapper class for the gRPC client inside (MyService). And yes, this worked with Quarkus 3.2.10.Final.

I managed to find a fix. Somewhere between Quarkus 3.2 and 3.8, a new config key was added: ["test-port"](https://quarkus.io/version/3.8/guides/grpc-service-consumption#quarkus-grpc-config-group-config-grpc-client-configuration_quarkus-grpc-clients-client-name-test-port) (can't find any documentation about it tho). Switching from:
`            System.setProperty("quarkus.grpc.clients.myService.**port**", port);`
to
`            System.setProperty("quarkus.grpc.clients.myService.**test-port**", port);`
fixed the problem.

Do you know why this config was introduced? There are already a few other ways to change the port between prod and test, why add yet another way to do it?

**Reply by cescoffier** [reply]
Oh, this was always there but it did not work because of a bug. In tests, you need test-port, which means you can have your gRPC service in dev mode and test mode simultaneously.
